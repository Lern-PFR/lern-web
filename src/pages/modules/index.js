import ModuleDetailsPage from './ModuleDetailsPage';

export {
	// eslint-disable-next-line import/prefer-default-export
	ModuleDetailsPage,
};
