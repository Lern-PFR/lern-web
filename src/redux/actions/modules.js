import * as ModulesApi from 'api/modulesApi';

/**
 * @constant
 * @name ActionTypes
 * @description The various action types used to interact with the modules redux state.
 * @type {object}
 *
 * @author Timothée Simon-Franza
 */
export const ActionTypes = {
	FETCH_MODULE_REQUEST: '@MODULES/FETCH_REQUEST',
	FETCH_MODULE_SUCCESS: '@MODULES/FETCH_SUCCESS',
	FETCH_MODULE_FAILURE: '@MODULES/FETCH_FAILURE',

	FETCH_MODULE_LIST_REQUEST: '@MODULES/FETCH_LIST_REQUEST',
	FETCH_MODULE_LIST_SUCCESS: '@MODULES/FETCH_LIST_SUCCESS',
	FETCH_MODULE_LIST_FAILURE: '@MODULES/FETCH_LIST_FAILURE',

	CREATE_MODULE_REQUEST: '@MODULES/CREATE_REQUEST',
	CREATE_MODULE_SUCCESS: '@MODULES/CREATE_SUCCESS',
	CREATE_MODULE_FAILURE: '@MODULES/CREATE_FAILURE',

	UPDATE_MODULE_REQUEST: '@MODULES/UPDATE_REQUEST',
	UPDATE_MODULE_SUCCESS: '@MODULES/UPDATE_SUCCESS',
	UPDATE_MODULE_FAILURE: '@MODULES/UPDATE_FAILURE',

	DELETE_MODULE_REQUEST: '@MODULES/DELETE_REQUEST',
	DELETE_MODULE_SUCCESS: '@MODULES/DELETE_SUCCESS',
	DELETE_MODULE_FAILURE: '@MODULES/DELETE_FAILURE',

	CLEAR_MODULE_LIST: '@MODULES/CLEAR_LIST',
};

// //////////////////////////////////////////////////////// //
// //////////// Specific module fetching actions ////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchModuleRequest
 * @description Action triggered anytime a module fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchModuleRequest = () => ({ type: ActionTypes.FETCH_MODULE_REQUEST });

/**
 * @function
 * @name fetchModuleSuccess
 * @description Action triggered as a result to a successful module fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} module : The module retrieved from the API.
 *
 * @returns {object}
 */
const fetchModuleSuccess = ({ module }) => ({
	type: ActionTypes.FETCH_MODULE_SUCCESS,
	payload: { module },
});

/**
 * @function
 * @name fetchModuleFailure
 * @description Action triggered as a result to a failed module fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchModuleFailure = (error) => ({
	type: ActionTypes.FETCH_MODULE_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// ////////////// Module list fetching actions //////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchModuleListRequest
 * @description Action triggered anytime a module list fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchModuleListRequest = () => ({ type: ActionTypes.FETCH_MODULE_LIST_REQUEST });

/**
 * @function
 * @name fetchModuleListSuccess
 * @description Action triggered as a result to a successful module list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {array}	modules		: The list of modules retrieved from the API.
 * @param {number}	totalCount	: The total amount of modules available for the subject.
 *
 * @returns {object}
 */
const fetchModuleListSuccess = ({ modules, totalCount }) => ({
	type: ActionTypes.FETCH_MODULE_LIST_SUCCESS,
	payload: { modules, totalCount },
});

/**
 * @function
 * @name fetchModuleListFailure
 * @description Action triggered as a result to a failed module list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchModuleListFailure = (error) => ({
	type: ActionTypes.FETCH_MODULE_LIST_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Module creation actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name createModuleRequest
 * @description Action triggered anytime a module creation call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const createModuleRequest = () => ({ type: ActionTypes.CREATE_MODULE_REQUEST });

/**
 * @function
 * @name createModuleSuccess
 * @description Action triggered as a result to a successfult module creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} module : The newly created module.
 *
 * @returns {object}
 */
const createModuleSuccess = ({ module }) => ({
	type: ActionTypes.CREATE_MODULE_SUCCESS,
	payload: { module },
});

/**
 * @function
 * @name createModuleFailure
 * @description Action triggered as a result to a failed module creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const createModuleFailure = (error) => ({
	type: ActionTypes.CREATE_MODULE_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Module edition actions //////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name updateModuleRequest
 * @description Action triggered anytime a module edition call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const updateModuleRequest = () => ({ type: ActionTypes.UPDATE_MODULE_REQUEST });

/**
 * @function
 * @name updateModuleSuccess
 * @description Action triggered as a result to a successful module edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} module : The updated module object.
 *
 * @returns {object}
 */
const updateModuleSuccess = ({ module }) => ({
	type: ActionTypes.UPDATE_MODULE_SUCCESS,
	payload: { module },
});

/**
 * @function
 * @name updateModuleFailure
 * @description Action triggered as a result to a failed module edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const updateModuleFailure = (error) => ({
	type: ActionTypes.UPDATE_MODULE_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Module deletion actions //////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name deleteModuleRequest
 * @description Action triggered anytime a module deletion call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const deleteModuleRequest = () => ({ type: ActionTypes.DELETE_MODULE_REQUEST });

/**
 * @function
 * @name deleteModuleSuccess
 * @description Action triggered as a result to a successful module deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} module : The deleted module object.
 *
 * @returns {object}
 */
const deleteModuleSuccess = ({ module }) => ({
	type: ActionTypes.DELETE_MODULE_SUCCESS,
	payload: { module },
});

/**
 * @function
 * @name deleteModuleFailure
 * @description Action triggered as a result to a failed module deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const deleteModuleFailure = (error) => ({
	type: ActionTypes.DELETE_MODULE_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Exported action creators ////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchModule
 * @description Method used to fetch the latest version of a specified module
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId : The id of the module we want to retrieve.
 */
export const fetchModule = (moduleId) => (dispatch) => {
	dispatch(fetchModuleRequest());

	return ModulesApi.fetchModuleById(moduleId)
		.then(({ module }) => dispatch(fetchModuleSuccess({ module })))
		.catch((error) => dispatch(fetchModuleFailure(error)));
};

/**
 * @function
 * @name fetchModuleListBySubjectId
 * @description Method used to fetch all modules from a specific subject, identified by the subjectId param.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} subjectId : The id of the subject we want to retrieve modules from.
 */
export const fetchModuleListBySubjectId = (subjectId) => (dispatch) => {
	dispatch(fetchModuleListRequest());

	return ModulesApi.fetchModulesBySubjectId(subjectId)
		.then(({ modules, totalCount }) => dispatch(fetchModuleListSuccess({ modules, totalCount })))
		.catch((error) => dispatch(fetchModuleListFailure(error)));
};

/**
 * @function
 * @name createModule
 * @description Method used to create a new module in the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} moduleData : The data to create the new module from.
 */
export const createModule = (moduleData) => (dispatch) => {
	dispatch(createModuleRequest());

	return ModulesApi.createModule(moduleData)
		.then(({ module }) => dispatch(createModuleSuccess({ module })))
		.catch((error) => dispatch(createModuleFailure(error)));
};

/**
 * @function
 * @name updateModule
 * @description Method used to update an existing module instance from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} moduleData	: The data to update the module with.
 * @param {string} moduleId		: The id to identify the module to update.
 */
export const updateModule = (moduleData, moduleId) => (dispatch) => {
	dispatch(updateModuleRequest());

	return ModulesApi.updateModule(moduleData, moduleId)
		.then(({ module }) => dispatch(updateModuleSuccess({ module })))
		.catch((error) => dispatch(updateModuleFailure(error)));
};

/**
 * @function
 * @name deleteModule
 * @description Method used to remove an existing module from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId : The id identifying the module to remove.
 */
export const deleteModule = (moduleId) => (dispatch) => {
	dispatch(deleteModuleRequest());

	return ModulesApi.deleteModule(moduleId)
		.then(({ module }) => dispatch(deleteModuleSuccess({ module })))
		.catch((error) => dispatch(deleteModuleFailure(error)));
};
