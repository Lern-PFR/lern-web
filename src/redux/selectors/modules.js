import { sortBy } from 'lodash';
import { getSubjectById } from './subjects';

const { createSelector } = require('reselect');

/**
 * @function
 * @name getModules
 * @description A selector callback which returns all modules from the Redux state.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} state The redux state at the time of call.
 *
 * @returns {Array}
 */
const getModules = (state) => state?.modules?.items ?? [];

/**
 * @function
 * @name getModuleById
 * @description A selector callback which returns the module whose id matched the moduleId parameter.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId The id of the module we want to retrieve.
 *
 * @returns {object|undefined}
 */
const getModuleById = createSelector(
	getModules,
	(_, moduleId) => moduleId,
	(modules, moduleId) => {
		const module = modules.filter(({ id }) => (id === moduleId))?.[0] ?? undefined;

		if (module === undefined) {
			return module;
		}

		return {
			...module,
			concepts: sortBy(module?.concepts || [], 'order'),
		};
	}
);

/**
 * @function
 * @name getModuleOrderOptions
 * @description A selector callback which returns an array of "order" values for the module edition form's order select field.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {Array}
 */
const getModuleOrderOptions = createSelector(
	getSubjectById,
	(subject) => {
		if (!subject?.modules?.length || subject?.modules?.length < 1) {
			return [{ label: 0, value: 0 }];
		}

		return Array.from({ length: subject?.modules?.length || 0 }, (_, i) => ({ label: i, value: i }));
	}
);

export {
	getModules,
	getModuleById,
	getModuleOrderOptions,
};
