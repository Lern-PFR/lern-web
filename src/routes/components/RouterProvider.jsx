import PropTypes from 'prop-types';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

/**
 * @name history
 * @description The global history object, handling the navigation for the whole application
 *
 * @author Yann Hodiesne
 */
export const history = createBrowserHistory();

/**
 * @name RouterProvider
 * @description A component providing a Router from react-router, with the ability to export its history and block navigation
 *
 * @author Yann Hodiesne
 *
 * @param {*} children : The children to render inside the Router
 */
const RouterProvider = ({ children }) => (
	<Router history={history}>
		{children}
	</Router>
);

RouterProvider.propTypes = {
	children: PropTypes.node.isRequired,
};

export default RouterProvider;
