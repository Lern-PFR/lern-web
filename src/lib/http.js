import localstorage from './localstorage';
import history from './history'
import { logoutAction } from '../redux/actions/auth';
import store from '../redux/store/configureStore';

const { dispatch, getState } = store;

/**
 * Generates HTTP headers.
 * 
 * If the user is authenticated, retrieves its jwt authentication token and add it to the header list.
 */
const getHeaders = () => {
    const headers = new Headers();
    const user = JSON.parse(localstorage.getItem('user'));
    
    headers.append('Content-Type', 'application/json');
  
    if (user && user.token) {
        headers.append('Authorization', `Bearer ${user.token}`);
    }
  
    return headers;
};

/**
 * Returns a query string from the object in parameter.
 * 
 * @param {object} params : The object to create the query string from.
 */
const objectToQS = (params) => Object
    .entries(params)
    .map(([key, value]) => {
        if (value !== undefined && value !== null) {
            return `${encodeURIComponent(key)}=${encodeURIComponent(value)}`;
        }

        return `${encodeURIComponent(key)}`;
    })
    .join('&');

/**
 * Generates a POST HTTP query from the url and params object in parameters.
 * 
 * @param {string} url      : The url to target.
 * @param {object} params   : The parameters to inject into the query.
 */
export const post = (url, params) => fetch(url, {
    method: 'POST',
    body: JSON.stringify(params),
    headers: getHeaders(),
});

/**
 * Generates a GET HTTP query from the url and params object in parameters.
 * 
 * If the params object is set, converts its content into a query string.
 * 
 * @param {string} url      : The url to target.
 * @param {object} params   : The parameters to inject into the query.
 */
export const get = (url, params) => {
    const qs = params ? `?${objectToQS(params)}` : '';

    return fetch(`${url}${qs}`, {
        headers: getHeaders(),
    });
};

/**
 * Generates a PUT HTTP query from the url and params object in parameters.
 * 
 * @param {string} url       : The url to target.
 * @param {object} params    : The parameters to inject into the query.
 */
export const put = (url, params) => fetch(url, {
    method: 'PUT',
    body: JSON.stringify(params),
    headers: getHeaders(),
});

/**
 * Generates a DELETE HTTP query from the url in parameters.
 * 
 * @param {string} url 
 */
export const del = (url) => fetch(url, {
    method: 'PUT',
    headers: getHeaders(),
});

/**
 * Handles the response from an HTTP query.
 * 
 * If a trigger is passed to the method, dispatch it using redux.
 * 
 * @param {*} trigger : The redux action to dispatch.
 */
export const handleResponse = (trigger) => (response) => {
    if (response.status === 204) {
      if (trigger) {
        dispatch(trigger());
      }
  
      return Promise.resolve();
    }
  
    if (response.status === 403 || response.status === 413 || response.status >= 500) {
      return Promise.reject(response);
    }
  
    return response.json().then((res) => {
        if (response.status === 401 && res.message !== 'Bad credentials') {
            dispatch(logoutAction());
            history.push('/');
    
            return Promise.reject({
                ...res,
                status: response.status,
            });
        }
    
        if (response.status >= 300) {
            return Promise.reject({
                ...res,
                status: response.status,
            });
        }
    
        if (trigger) {
            dispatch(trigger(res));
        }
    
        return Promise.resolve(res);
    });
};