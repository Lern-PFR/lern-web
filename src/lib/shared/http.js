import { history } from 'routes/components/RouterProvider';
import session from 'lib/shared/session';
import routes from 'routes';

/**
 * @constant
 * @name baseUrl
 * @description The base url to append to in order to target and API endpoint.
 * @type {string}
 *
 * @author Timothée Simon-Franza
 *
 * Note : Its value is set using environment variables set defined in this .env file and forwarded by docker-compose.
 */
export const baseUrl = process.env.REACT_APP_API_URL;

/**
 * @function
 * @name getHeaders
 * @description Generates and returns a header bundle from the current session's data.
 *
 * @author Timothée Simon-Franza
 * @author Yann Hodiesne
 *
 * @returns {Header} the generated header bundle.
 */
export const getHeaders = () => {
	const headers = new Headers();

	headers.append('Content-Type', 'application/json');

	if (session.exists()) {
		headers.append('Authorization', `Bearer ${session.get()}`);
	}

	return headers;
};

/**
 * @function
 * @name objectToQS
 * @description Converts the object in parameter to a query string.
 *
 * @author Timothée Simon-Franza
 * @author Yann Hodiesne
 *
 * @param {object} params : The object to convert to QS.
 *
 * @returns {string}
 */
export const objectToQS = (params) => Object
	.entries(params)
	.map(([key, value]) => {
		if (value !== undefined && value !== null) {
			return `${encodeURIComponent(key)}=${encodeURIComponent(value)}`;
		}

		return `${encodeURIComponent(key)}`;
	})
	.join('&');

/**
 * @function
 * @name handleResponse
 * @description Automatically deserializes and handles the received response object.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} response : The response to handle.
 *
 * @returns {Promise | object}
 */
export const handleResponse = (response) => {
	if (response.status === 204) {
		return Promise.resolve();
	}

	if (response.status === 401 && session.exists()) {
		history.push(routes.auth.logout);
	}

	if (response.status === 401 || response.status === 403 || response.status === 409 || response.status === 413 || response.status >= 500) {
		return Promise.reject(response);
	}

	return response.json();
};

/**
 * @function
 * @name handleError
 * @description Automatically handles the received error to return a formatted object.
 *
 * @author Yann Hodiesne
 *
 * @param {object} error : The error to handle.
 *
 * @returns {Promise.reject}
 */
export const handleError = (error) => {
	const {
		status,
		statusText: message,
	} = error;

	return Promise.reject({ status, message });
};

/**
 * @function
 * @name post
 * @description Creates a POST method http call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} url	: The url to target.
 * @param {object} body	: The body to include.
 *
 * @returns {Promise | object}
 */
export const post = (url, body) => fetch(`${baseUrl}${url}`, {
	method: 'POST',
	body: JSON.stringify(body),
	headers: getHeaders(),
}).then(handleResponse).catch(handleError);

/**
 * @function
 * @name get
 * @description Creates a GET method http call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} url		: The url to target.
 * @param {objcet} qsObject	: The object to convert into a query string.
 *
 * @returns {Promise | object}
 */
export const get = (url, qsObject) => {
	const queryString = (qsObject && Object.keys(qsObject).length !== 0) ? `?${objectToQS(qsObject)}` : '';

	return fetch(`${baseUrl}${url}${queryString}`, {
		method: 'GET',
		headers: getHeaders(),
	}).then(handleResponse).catch(handleError);
};

/**
 * @function
 * @name put
 * @description Creates a PUT method http call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} url : The url to target.
 * @param {object} body : The body to include.
 */
export const put = (url, body) => fetch(`${baseUrl}${url}`, {
	method: 'PUT',
	body: JSON.stringify(body),
	headers: getHeaders(),
}).then(handleResponse).catch(handleError);

/**
 * @function
 * @name del
 * @description Creates a DELETE method http call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} url : The url to target.
 */
export const del = (url) => fetch(`${baseUrl}${url}`, {
	method: 'DELETE',
	headers: getHeaders(),
}).then(handleResponse).catch(handleError);
