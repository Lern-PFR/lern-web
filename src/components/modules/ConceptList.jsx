import PropTypes from 'prop-types';
import { StyledList } from 'components/shared/styledElements';
import { conceptList } from 'theme/pages/modules/moduleDetailsPage';
import ConceptListItem from './ConceptListItem';

/**
 * @name ConceptList
 * @description A component used to display a module's concept list.
 *
 * @author TimothÃ©e Simon-Franza
 *
 * @param {array} concepts The list of concepts to display.
 */
const ConceptList = ({ concepts }) => (
	<StyledList listStyle="none" {...conceptList}>
		{concepts && concepts.map(({ id, title, description, courses }) => (
			<ConceptListItem key={id} id={id} title={title} description={description} lessons={courses} />
		))}
	</StyledList>
);

ConceptList.propTypes = {
	concepts: PropTypes.arrayOf(PropTypes.shape({
		id: PropTypes.string.isRequired,
		title: PropTypes.string.isRequired,
		description: PropTypes.string.isRequired,
		lessons: PropTypes.arrayOf(PropTypes.shape({
			id: PropTypes.string.isRequired,
			name: PropTypes.string.isRequired,
			description: PropTypes.string.isRequired,
		})),
	})).isRequired,
};

export default ConceptList;
