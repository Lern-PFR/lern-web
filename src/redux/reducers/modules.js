import { ActionTypes } from 'redux/actions/modules';

const initialState = {
	isLoading: false,
	items: [],
	totalCount: 0,
};

export default (state = initialState, { type, payload }) => {
	switch (type) {
		case ActionTypes.FETCH_MODULE_REQUEST:
		case ActionTypes.FETCH_MODULE_LIST_REQUEST:
		case ActionTypes.CREATE_MODULE_REQUEST:
		case ActionTypes.UPDATE_MODULE_REQUEST:
		case ActionTypes.DELETE_MODULE_REQUEST:
			return {
				...state,
				isLoading: true,
			};
		case ActionTypes.FETCH_MODULE_FAILURE:
		case ActionTypes.FETCH_MODULE_LIST_FAILURE:
		case ActionTypes.CREATE_MODULE_FAILURE:
		case ActionTypes.UPDATE_MODULE_FAILURE:
		case ActionTypes.DELETE_MODULE_FAILURE:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.CREATE_MODULE_SUCCESS:
		case ActionTypes.UPDATE_MODULE_SUCCESS:
		case ActionTypes.DELETE_MODULE_SUCCESS:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.FETCH_MODULE_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [payload.module],
				totalCount: 1,
			};
		case ActionTypes.FETCH_MODULE_LIST_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [...payload.modules],
				totalCount: payload.totalCount,
			};
		case ActionTypes.CLEAR_MODULE_LIST:
			return {
				...state,
				items: [],
				totalCount: 0,
			};
		default:
			return state;
	}
};
