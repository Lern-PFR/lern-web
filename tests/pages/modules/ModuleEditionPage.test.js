import { mount, shallow } from 'enzyme';
import { ModuleEditionPage } from 'pages/modules';

import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import routeData from 'react-router';
import { baseUrl } from 'lib/shared/http';

import RouterProvider from 'routes/components/RouterProvider';
import { ActionTypes as ModulesActionTypes } from 'redux/actions/modules';
import fetchMock from 'fetch-mock';

const mockStore = configureMockStore([thunk]);

describe('Module edition page', () => {
	let store;

	beforeEach(() => {
		jest.spyOn(routeData, 'useParams').mockReturnValue({ moduleId: 'dummy_module_id' });
	});

	afterEach(() => {
		jest.restoreAllMocks();
		jest.resetAllMocks();
	});

	describe('Snapshot testing', () => {
		it('should match previous snapshot without the required module in state', () => {
			store = mockStore({});

			const sut = (
				<Provider store={store}>
					<RouterProvider>
						<ModuleEditionPage />
					</RouterProvider>
				</Provider>
			);

			const wrapper = shallow(sut);

			expect(wrapper).toMatchSnapshot();
		});

		it('should match previous snapshot with the required module and subject in state', () => {
			store = mockStore({
				subjects: {
					all: [{
						id: 'dummy_subject_id',
						title: 'dummy_subject_title',
						description: 'dummy_subject_desc',
						modules: [{
							id: 'dummy_module_id',
							title: 'dummy_module_title',
							description: 'dummy_subject_desc',
							subjectId: 'dummy_subject_id',
							order: 0,
						}],
					}],
				},
			});

			const sut = (
				<Provider store={store}>
					<RouterProvider>
						<ModuleEditionPage />
					</RouterProvider>
				</Provider>
			);

			const wrapper = shallow(sut);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('dispatched method calls', () => {
		it('should call the fetchModule action creator on mount', async () => {
			const subjectHttpResponse = {
				status: 200,
				body: { id: 'dummy_subject_id', name: 'subject_name' },
				headers: { 'content-type': 'application/json' },
			};

			const moduleHttpResponse = {
				status: 200,
				body: { id: 'dummy_module_id', name: 'modulet_name', subjectId: 'dummy_subject_id' },
				headers: { 'content-type': 'application/json' },
			};

			fetchMock.get(`${baseUrl}/api/modules/dummy_module_id`, moduleHttpResponse);
			fetchMock.get(`${baseUrl}/api/subjects/dummy_subject_id`, subjectHttpResponse);

			mount(
				<Provider store={store}>
					<RouterProvider>
						<ModuleEditionPage />
					</RouterProvider>
				</Provider>
			);

			const expectedActions = [
				{ type: ModulesActionTypes.FETCH_MODULE_REQUEST },
			];

			expect(store.getActions()).toEqual(expectedActions);
		});
	});
});
