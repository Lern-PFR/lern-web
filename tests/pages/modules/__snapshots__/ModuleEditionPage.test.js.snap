// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module edition page Snapshot testing should match previous snapshot with the required module and subject in state 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <RouterProvider>
    <ModuleEditionPage />
  </RouterProvider>
</ContextProvider>
`;

exports[`Module edition page Snapshot testing should match previous snapshot without the required module in state 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <RouterProvider>
    <ModuleEditionPage />
  </RouterProvider>
</ContextProvider>
`;
